src/typegenTypeHelpers.ts

```diff
@@ -255,3 +260,221 @@ export type NeedsResolver<TypeName extends string, FieldName extends string> = H
+  ? {
+      /**
+       * Optionally provide a custom type resolver function. If one is not provided,
+       * the default implementation will call `isTypeOf` on each implementing
+       * Object type.
+       */
+      resolveType?: AbstractTypeResolver<TypeName>
+    }
+  : {
+      /**
+       * Optionally provide a custom type resolver function. If one is not provided,
+       * the default implementation will call `isTypeOf` on each implementing
+       * Object type.
+       */
+      resolveType: AbstractTypeResolver<TypeName>
+    }
+
+/**
+ * Returns whether a field which type is either an 'Object | Interface | Enum' should discriminate its return type with __typename
+ */
+export type ShouldDiscriminateOutputTypeField<OutputTypeName extends string> = OutputTypeName extends GetGen<
+  'isTypeOfObjectNames'
+> // if it implements isTypeOf already
+  ? false // then don't disable __typename feature
+  : ResolveTypeImplementedInAllAbstractTypes<OutputTypeName> extends true // else if abstract implement resolve type
+  ? false // then disable __typename feature
+  : true // otherwise enable __typename feature
+
+export type ShouldDiscriminateAbstractTypeField<
+  AbstractTypeName extends string
+> = AbstractTypeName extends GetGen<'resolveTypeImplemented'> // if the abstract type has resolve type already
+  ? false // then disable __typename feature
+  : IsTypeOfImplementedInAllMembers<AbstractTypeName> extends true // if all members of the abstract type implements isTypeOf
+  ? false // then disable __typename feature
+  : true // otherwise enable __typename feature
+
+/**
+ * Returns whether a field which type is a union should discriminate its return type with __typename
+ */
+export type ShouldDiscriminateResultValue<
+  TypeName extends string,
+  FieldName extends string,
+  ReturnTypeName extends string = GetGen3<'fieldTypeNames', TypeName, FieldName>
+> = GetGen3<'features', 'abstractTypes', '__typename', false> extends true // if feature is enabled
+  ? ReturnTypeName extends GetGen<'objectNames'> | GetGen<'interfaceNames'> | GetGen<'enumNames'> // and return type is either an object type || interface || enum
+    ? ShouldDiscriminateOutputTypeField<ReturnTypeName> extends true // call sub function
+      ? true
+      : false
+    : ReturnTypeName extends GetGen<'abstractTypes'> // else if return type is an abstract type
+    ? ShouldDiscriminateAbstractTypeField<ReturnTypeName> extends true // call sub function
+      ? true
+      : false
+    : false
+  : false
+
+/**
+ * Discriminate a type with a __typename: TypeName field
+ */
+type DiscriminateImpl<TypeName extends string, Type> = Type extends Promise<infer A>
+  ? Promise<DiscriminateImpl<TypeName, A>>
+  : Type extends Array<infer B>
+  ? Array<DiscriminateImpl<TypeName, B>>
+  : Type extends ReadonlyArray<infer C>
+  ? ReadonlyArray<DiscriminateImpl<TypeName, C>>
+  : Type extends object
+  ? Type & { __typename: TypeName }
```

Comment:
IIRC there is somewhere else in the codebase we do similar or probably same unwrapping of types to get at the actual returned model type. I believe this is pretty fragile logic and should be centralized. If we are duplicating this logic think it might be interesting to have a helper like: 

```ts
type DiscriminateImpl<ModelType, TypeName extends string> = IntersectModelTypeWith<ModelType, { __typename: TypeName }>

type IntersectModelTypeWith<ModelType, T> = ...all the conditional recursive logic... finally... ModelType & T
```

This also helps separate the "business logic" from the lower-level details about how to drill down to the model type. Thus in the end this is a lot easier to read/declarative: `IntersectModelTypeWith<ModelType, { __typename: TypeName }>` IMO